pkg/controller/certificates/certificate_request.go

```diff
@@ -0,0 +1,916 @@
+			secretExists = false
+		}
+	}
+	reqs, err := findCertificateRequestsForCertificate(log, crt, c.certificateRequestLister)
+	if err != nil {
+		return err
+	}
+	var req *cmapi.CertificateRequest
+	if len(reqs) == 1 {
+		req = reqs[0]
+	}
+	var cert *x509.Certificate
+	var certExpired bool
+	if len(certs) > 0 {
+		cert = certs[0]
+		certExpired = cert.NotAfter.Before(c.clock.Now())
+	}
+
+	var matches bool
+	var matchErrs []string
+	if key != nil && cert != nil {
+		matches, matchErrs = certificateMatchesSpec(crt, key, cert, c.secretLister)
+	}
+
+	isTempCert := isTemporaryCertificate(cert)
+
+	// begin setting certificate status fields
+	if !matches || isTempCert {
+		crt.Status.NotAfter = nil
+	} else {
+		metaNotAfter := metav1.NewTime(cert.NotAfter)
+		crt.Status.NotAfter = &metaNotAfter
+	}
+
+	// Derive & set 'Ready' condition on Certificate resource
+	ready := cmapi.ConditionFalse
+	reason := ""
+	message := ""
+	switch {
+	case !secretExists || key == nil:
+		reason = "NotFound"
+		message = "Certificate does not exist"
+	case matches && !isTempCert && !certExpired:
+		ready = cmapi.ConditionTrue
+		reason = "Ready"
+		message = "Certificate is up to date and has not expired"
+	case req != nil:
+		reason = "InProgress"
+		message = fmt.Sprintf("Waiting for CertificateRequest %q to complete", req.Name)
+	case cert == nil:
+		reason = "Pending"
+		message = "Certificate pending issuance"
+	case !matches:
+		reason = "DoesNotMatch"
+		message = strings.Join(matchErrs, ", ")
+	case certExpired:
+		reason = "Expired"
+		message = fmt.Sprintf("Certificate has expired on %s", cert.NotAfter.Format(time.RFC822))
+	case isTempCert:
+		reason = "TemporaryCertificate"
+		message = "Certificate issuance in progress. Temporary certificate issued."
+	default:
+		// theoretically, it should not be possible to reach this state.
+		// practically, we may have missed some edge cases above.
+		// print a dump of the current state as a log message so that users can
+		// discover, share and attempt to resolve bugs in this area of code easily.
+		log.Info("unknown certificate state",
+			"secret_exists", secretExists,
+			"matches", matches,
+			"is_temp_cert", isTempCert,
+			"cert_expired", certExpired,
+			"key_is_nil", key == nil,
+			"req_is_nil", req == nil,
+			"cert_is_nil", cert == nil,
+		)
+		ready = cmapi.ConditionFalse
+		reason = "Unknown"
+		message = "Unknown certificate status. Please open an issue and share your controller logs."
+	}
+	apiutil.SetCertificateCondition(crt, cmapi.CertificateConditionReady, ready, reason, message)
+
+	_, err = updateCertificateStatus(ctx, c.cmClient, old, crt)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// processCertificate is the core method that is called in the manager.
+// It accepts a Certificate resource, and checks to see if the certificate
+// requires re-issuance.
+func (c *certificateRequestManager) processCertificate(ctx context.Context, crt *cmapi.Certificate) error {
```

Comment:
Not a huge fan of how big this function is. Would it be better to split it up into stages?